Fix unable to remove empty blocks on merge

kevin940726 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What and why?
<!-- In a few words, what is the PR actually doing? -->
Fix https://github.com/WordPress/gutenberg/issues/65174.

Arguably considered intended, but it could be clearer for users when trying to delete an empty block on merge.

This PR prevents empty unmodified blocks from being moved to the parent on merge. Modified blocks can still be moved to the parent to retain backward compatibility and user expectations.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Refactor and reorder the `moveFirstItemUp` implementation. Now it follows the following rules:

1. Transform the block to a default block if it's empty and can be transformed.
2. Remove the block if the block is already a default block and empty.
3. Move the block up to the parent if it can be moved.
4. Move the transformed block up to the parent if the block can be transformed to a default block and can be inserted at the new location.
5. Continue the fallback behavior:
  1. Transform the block to a default block if it can be transformed
  2. Remove the block if it's an unmodified default block

An "empty" block is defined as if the `content` attributes are empty.

One use case for these complicated rules is to handle the empty list item v.s. empty heading.

In https://github.com/WordPress/gutenberg/issues/65174, an empty heading is expected to be removed on backspace, but an empty list item should be converted to a default block and lifted to the parent. Both blocks are "empty" and not default blocks. The only difference is that an empty list item cannot be transformed to a default block inline, while an empty heading block can. So we try converting the empty heading block to an empty paragraph block on the first backspace (instead of removing it) to provide users instant feedback on backspace. I think it's an acceptable trade-off but there could be other better solutions!

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

Added e2e tests `should handle unwrapping and merging blocks`.

Follow the testing instructions in the original issue: https://github.com/WordPress/gutenberg/issues/65174.

Or, you can try copy-paste this markup and try interacting with it.
<details><summary>Code</summary>
<p>

```html
<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"align":"center"} -->
<p class="has-text-align-center"></p>
<!-- /wp:paragraph -->

<!-- wp:heading {"textAlign":"center"} -->
<h2 class="wp-block-heading has-text-align-center"></h2>
<!-- /wp:heading -->

<!-- wp:heading -->
<h2 class="wp-block-heading">Heading</h2>
<!-- /wp:heading -->

<!-- wp:separator -->
<hr class="wp-block-separator has-alpha-channel-opacity"/>
<!-- /wp:separator --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->
```

</p>
</details>

Otherwise, follow the steps below:

1. Open the post editor.
3. Insert a group block, and insert an empty paragraph in the group block. Insert another arbitrary block at the end of the group block.
4. Focus on the empty block and press <kbd>Backspace</kbd>
5. Expect the block to be deleted rather than moved up.
7. Try the same setup again with a heading block and expect it to be moved to parent and converted to a paragraph block.
8. Try other combinations.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
Same as above.

## Screenshots or screencast <!-- if applicable -->


https://github.com/user-attachments/assets/1d0b88f1-e150-4038-a09b-cfbed64b6e62

kevin940726 committed:
Fix unable to remove empty blocks on merge

kevin940726 commented on test/e2e/specs/editor/blocks/list.spec.js:
```diff
@@ -1237,17 +1237,17 @@ test.describe( 'List (@firefox)', () => {
-<!-- /wp:list-item --></ul>
-<!-- /wp:list -->`
-		);
+		await expect.poll( editor.getBlocks ).toMatchObject( [
```
This test seems to have been broken since #42711.

Now it matches the result before the migration:

https://github.com/WordPress/gutenberg/pull/41555/files#diff-97ba09329f315d385933ebd76613f9f2f053ad71d1fb3812cc16b6f7d4fd40aaL83-L87

However, in #42711, it was changed to two paragraphs:

https://github.com/WordPress/gutenberg/pull/42711/files#diff-2d61608781bfdb7c9aea282172506baca36e1c8307f774f3f390328ccee02c67R1066-R1068.

In #45075, it was fixed, but only partially:

https://github.com/WordPress/gutenberg/pull/45075/files#diff-2d61608781bfdb7c9aea282172506baca36e1c8307f774f3f390328ccee02c67L1093-R1097

Since the test has survived for two changes in different PRs, I'm not sure what's the expected behavior anymore 😅. I think the current result makes sense, but I could use some double-checking from past reviewers too. 🙏

kevin940726 replied:
It seems like the first empty paragraph is actually expected to be there! I've updated the test.

ntsekouras commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -329,6 +330,24 @@ const applyWithDispatch = withDispatch( ( dispatch, ownProps, registry ) => {
+			 * @param {string} blockClientId The block client ID.
+			 * @return {boolean} Whether the block content is empty.
+			 */
+			function isBlockContentEmpty( blockClientId ) {
```
Why would we need this function instead of just using `isUnmodifiedBlock`?

I was curious if the fix could be simpler and opened [a draft PR](https://github.com/WordPress/gutenberg/pull/65266) to check the CI..

kevin940726 replied:
This is to handle some blocks that are modified but still considered "empty". For instance, a heading block with some attributes set but without the content. There's also an added test case in this PR.

I'm not entirely sure if this is the intended way though. We can probably define another "rule" to fix the same thing in https://github.com/WordPress/gutenberg/issues/65174.

ntsekouras replied:
Okay, got it. Thank you.

talldan commented:
Unfortunately it seems this fix breaks the feature introduced in https://github.com/WordPress/gutenberg/pull/45075. If I had to pick one, I feel like it'd be better to keep the behavior in trunk, the way it works for lists is pretty useful.

I'm not sure if there's a way to keep the behavior for lists (and maybe quotes) but delete blocks like headings and paragraphs in groups.

The code here is very generic, not really reasoning about block types, so it might be tricky.

kevin940726 commented:
> Unfortunately it seems this fix breaks the feature introduced in https://github.com/WordPress/gutenberg/pull/45075.

I think unwrapping list items still works from my testing and the e2e tests? Could you point me to the features that are breaking after this PR?

andrewserong commented:
> Unfortunately it seems this fix breaks the feature introduced in https://github.com/WordPress/gutenberg/pull/45075. If I had to pick one, I feel like it'd be better to keep the behavior in trunk, the way it works for lists is pretty useful.

I was a bit stumped trying to reproduce this at first, as this PR mostly feels quite good to me in testing. As far as I could tell the main difference for lists appears to be what happens when you press backspace from an empty list item.

I.e. in this PR, the overall list becomes selected:

https://github.com/user-attachments/assets/d08662b9-0644-45b5-b1f5-eba5a9d1fd4c

Whereas on `trunk`, if we hit backspace from that first list item, we get an empty paragraph block. I think that's probably the desired behaviour so that it's easy to continue writing in a paragraph?

https://github.com/user-attachments/assets/62b47d69-f5a7-4152-894f-92d735455c7e

> I'm not sure if there's a way to keep the behavior for lists (and maybe quotes) but delete blocks like headings and paragraphs in groups.

I'm not sure how we'd do it either, but I tend to agree that it'd be ideal if we could have it working as in this PR for headings and paragraphs in groups, but retain the behaviour in `trunk` for lists.

Were there any other details from https://github.com/WordPress/gutenberg/pull/45075 that aren't working in this PR @talldan?

ntsekouras commented on packages/blocks/src/api/utils.js:
```diff
@@ -33,31 +33,38 @@ const ICON_COLORS = [ '#191e23', '#f8f9f9' ];
-			if ( definition.hasOwnProperty( 'default' ) ) {
-				return value === definition.default;
-			}
+export function isUnmodifiedBlock( block, attributeKeys ) {
```
Might it better to leave `isUnmodifiedBlock` intact and create a new util with our implementation of whether a block can be considered empty?

Additionally, should the empty block function just check for

```
if ( definition.type === 'rich-text' ) {
	return ! value?.length;
}
```

Do you see any drawbacks on that?

kevin940726 replied:
Third-party blocks could have content attributes that are not rich-text, (but would it able to call `onMerge`?). I think it's better to keep it a little bit generic for now. 😅

talldan replied:
I think many third party blocks are also unlikely to have `__experimentalRole` defined because of the way it's experimental.

getdave replied:
Just FYI it's been stablised now but Dan's point still stands:

https://github.com/WordPress/gutenberg/pull/65484

kevin940726 replied:
Yeah, I'm just trying to be safe and future-compatible here. (On the other hand, an attribute could also have the `rich-text` type but not be considered content.)

Thanks for the link! I can rebase this to use the stabilized function.

kevin940726 commented:
I updated the PR to include some more testing cases and rules. This is getting a lot more complicated though. Some feedback on the desired behavior is appreciated! I applied some predefined rules that I think are less frustrating to regular users but also tried to maintain maximum backward compatibility. See the recordings in the PR descriptions for some edge cases.

andrewserong commented:
Thanks for the updates, Kai!

> This is getting a lot more complicated though. Some feedback on the desired behavior is appreciated!

The behaviour for the first item in a List feels good to me know (conversion to the default block), same with an empty Paragraph block within a Group.

However, with the Heading block, I'd expect hitting backspace on an empty Heading block within a Group block to work the same as for the paragraph block — that is, for it to delete the Heading block. In the screengrabs and in local testing, it looks like it now moves up a level and converts to a default block, similar to the List item, which felt a bit unexpected to me:

https://github.com/user-attachments/assets/7fd7dc7d-6716-4e3c-92d0-f241884a6fe9

In the above video:

* Deleting the Paragraph block works as expected
* Deleting the Heading block moves up a level and converts to an empty paragraph block which felt unexpected to me

Unfortunately I can't quite think of a good way to handle this, but it seems to me like the behaviour of headings and paragraphs should probably be much the same as each other, but the behaviour for list items is the outlier? Are there any other primitive blocks we'd need to consider here, or is it mostly about paragraphs and headings that we want to deal with? 

Just wondering if it'd help to make it about the specific blocks we're dealing with, rather than keeping the logic generic, or would that only create further complexity?

ramonjd commented:
I can confirm that the base case of deleting an empty block (and that block being removed) is working on this PR:

### Trunk

https://github.com/user-attachments/assets/b1952924-2661-4461-8813-2f1eb18baf4d

### This PR


https://github.com/user-attachments/assets/d1e71667-9093-442d-adf0-727d1cc49e9b


> Deleting the Heading block moves up a level and converts to an empty paragraph block which felt unexpected to me

I would expect the empty block to be deleted in all case.

But, in general, think I failed the IQ test - testing this stuff seems very subtle and I'm not sure what's right or wrong anymore 🤣 For example, not related to this PR, but it's confusing for me how subsequent blocks are pushed outside the Group block when I create new empty blocks.

https://github.com/user-attachments/assets/16e12ee1-55ea-4964-b7d8-7570b7f4ee6b

kevin940726 commented:
> However, with the Heading block, I'd expect hitting backspace on an empty Heading block within a Group block to work the same as for the paragraph block — that is, for it to delete the Heading block.

Yeah, I think we have a few options, I'm not sure which one is the best TBH.

1. Delete it
2. Convert to an empty paragraph block

I think both solutions make sense in a way. The first solution might be the most straightforward but a little bit destructive. The second solution conveys the meaning of "resetting formatting" when pressing backspace on an empty block. This kind of design language can be seen in some rich text editors, but very subtle. It also matches the behavior of pressing backspace on a non-empty block more closely too. That said, I feel it might not be worth it at this stage, so I'll try the first approach first 😅.

kevin940726 commented:
I updated the PR yet again to hopefully handle the list item use case, too 😅. I also updated the PR description, but I'll copy-paste it here for visibility.

Now it follows the following rules:

1. Transform the block to a default block if it's empty and can be transformed.
2. Remove the block if the block is already a default block and empty.
3. Move the block up to the parent if it can be moved.
4. Move the transformed block up to the parent if the block can be transformed to a default block and can be inserted at the new location.
5. Continue the fallback behavior:
  1. Transform the block to a default block if it can be transformed
  2. Remove the block if it's an unmodified default block

An "empty" block is defined as if the `content` attributes are empty.

One use case for these complicated rules is to handle the empty list item v.s. empty heading.

In https://github.com/WordPress/gutenberg/issues/65174, an empty heading is expected to be removed on backspace, but an empty list item should be converted to a default block and lifted to the parent. Both blocks are "empty" and not default blocks. The only difference is that an empty list item cannot be transformed to a default block inline, while an empty heading block can. So we try converting the empty heading block to an empty paragraph block on the first backspace (instead of removing it) to provide users instant feedback on backspace. I think it's an acceptable trade-off but there could be other better solutions!

https://github.com/user-attachments/assets/1d0b88f1-e150-4038-a09b-cfbed64b6e62

ramonjd commented:
Tested with paragraph and heading blocks:

1. Empty blocks are removed as expected when hitting backspace
2. Non-empty blocks are moved up to parent as expected

I can't really see anything strange with list blocks so far, but like I said - there are many subtleties to this behaviour I'm not completely covering.

However in my testing this PR satisfies the reported issue: https://github.com/WordPress/gutenberg/issues/65174

andrewserong reviewed (commented):
> Tested with paragraph and heading blocks:

Same, this is testing well for me with headings, paragraphs, and lists 👍. Since it is a little nuanced this one, I'll just ping @WordPress/gutenberg-core for visibility in case anyone has any concerns about the proposed change here.

youknowriad commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -307,6 +311,26 @@ const applyWithDispatch = withDispatch( ( dispatch, ownProps, registry ) => {
+			 * @param {WPBlock} block The block to check.
+			 * @return {boolean} Whether the block is empty.
+			 */
+			function isBlockEmpty( block ) {
```
How is this different to the existing `isUnmodifiedBlock` selector that we have?

kevin940726 replied:
It's also discussed here: https://github.com/WordPress/gutenberg/pull/65262#discussion_r1756466550.

An "empty" block could still have some attributes modified. For instance, a heading block could have its content empty but also aligned to the center. The intent is to assume that the user might want to remove/reset the block when it's empty instead of moving it up to the parent when it has contents. Note that this is not the only solution but a proposed trade-off to fix https://github.com/WordPress/gutenberg/issues/65174.

youknowriad replied:
So an "empty" block is a block whose "content" role attribute are unmodified. 

I think this function should be removed `isBlockContentUnmodified` and probably be moved to `blocks` package as a private selector maybe, that way it's placed closed to the other `isUnmodifiedBlock`.

Now for the behavior itself, I understand the tradeoff, I think it's an ok tradeoff if this flow only applies to "text" blocks which is probably true. But I'm concerned about the impact on third-party "text" blocks because we just stabilized "content" role and I'm sure no one is actually using it. To make sure I understand the tradeoff better, what happens if a third-party text block is not assigning `role="content"` properly to its attribute?

kevin940726 replied:
> I think this function should be removed `isBlockContentUnmodified` and probably be moved to `blocks` package as a private selector maybe, that way it's placed closed to the other `isUnmodifiedBlock`.

This PR also adds a private `isAttributeUnmodified` function in the `blocks` package before the content role is stabilized. Since that it's now stabilized, I think we can move (and rename) `isBlockEmpty` there too.

> what happens if a third-party text block is not assigning `role="content"` properly to its attribute?

If there's no `content` role attributes, then it's also considered "empty". Then it'll follow the same steps as mentioned above.

youknowriad replied:
>If there's no content role attributes, then it's also considered "empty". Then it'll follow the same steps as mentioned above.

So content loss? I guess there's always "undo" but I wonder if we can find a better way here.

youknowriad replied:
I'm wondering if something like this would work better:

 - If a block has "content" attribute, use `isBlockContentUnmodified`, if not fallback to `isBlockUnmodified`

kevin940726 replied:
I think that's a fair trade-off for now. I just rebased and updated the branch!

kevin940726 committed:
Update to the stabilized API

youknowriad commented:
Would appreciate an approval from someone else as I didn't get to test this deeply but I do appreciate the changes and I think it's looking good. Thanks

Mamaduka commented on packages/blocks/src/api/utils.js:
```diff
@@ -30,6 +30,30 @@ extend( [ namesPlugin, a11yPlugin ] );
+ * @param {*}      value               The attribute's value.
+ * @return {boolean} Whether the attribute is unmodified.
+ */
+function isAttributeUnmodified( attributeDefinition, value ) {
```
Nit: Maybe we should follow naming conventions of similar methods - `isUnmodifiedAttribute`.

Mamaduka commented on packages/blocks/src/api/utils.js:
```diff
@@ -73,6 +84,35 @@ export function isUnmodifiedDefaultBlock( block ) {
+ * @param {WPBlock} block Block Object
+ * @return {boolean} Whether the block content is unmodified.
+ */
+export function isBlockContentUnmodified( block ) {
```
Nit (same): `isUnmodifiedBlockContent`

Mamaduka reviewed (approved):
This is testing well for me. As far as I remember, the e2e test coverage is pretty good for this area.

I think it's better to ship this a follow-up with bug fixes if they're discovered during the beta/RC cycle.

kevin940726 committed:
Rename the utils
